File: src/edu/stmc/SPRT.java
package edu.stmc;

/** Sequential Probability Ratio Test */
public class SPRT {
  private SPRT() { }

  /**
   * SPRT for binary simple hypothesis testing.
   * <ul>
   * <li><em>Binary</em> means if the algorithm terminates, there are two possibilities for its answer.
   * <li><em>Simple</em> means null and alternative hypotheses must be in the form of {@literal H0: p=p0} and
   * {@literal H1: p=p1}, respectively (for some values of {@code 0<p0<p1<1}). Note that {@literal H0: p<p0} is
   * not a simple hypothesis.</li>
   * </ul>
   */
  public static final class Binary implements HypTest {

    /**
     * Construct an instance of the binary version of SPRT
     * @param threshold Input threshold
     * @param alpha     Type I error probability (also known as <em>false positive</em> probability) is the probability
     *                  of incorrectly rejecting the null hypothesis.
     * @param beta      Type II error probability (also known as <em>false negative</em> probability) is the probability
     *                  of incorrectly not rejecting the null hypothesis.
     * @param delta     Half of the size of indifference region.
     * @requires {@code 0 < lb < threshold < ub < ub < 1}
     * @requires {@code 0 < alpha < 0.5}
     * @requires {@code 0 < beta < 0.5}
     */
    public Binary(final double threshold, final double alpha, final double beta, final double delta) {
      assert 0 < threshold && threshold < 1 : "Wrong threshold " + threshold;
      assert 0 < alpha && alpha < 0.5 : "Wrong type I error " + alpha;
      assert 0 < beta && beta < 0.5 : "Wrong type II error " + beta;
      assert delta < threshold && threshold < 1 - delta : "Invalid threshold (" + threshold + ") and/or delta (" + delta + ")";
      assert 0 < alpha && alpha < 0.5 : "Wrong type I error " + alpha;
      assert 0 < beta && beta < 0.5 : "Wrong type II error " + beta;
      final double lb = threshold - delta;
      final double ub = threshold + delta;
      this.threshold = threshold;
      this.alpha = alpha;
      this.beta = beta;
      this.delta = delta;
      logL = Math.log(beta / (1 - alpha));
      logU = Math.log((1 - beta) / alpha);
      assert logL < 0 : "Lower-bound log is not negative " + logL;
      assert logU > 0 : "Upper-bound log is not negative " + logU;
      q0 = Math.log((1 - ub) / (1 - lb));
      q1 = Math.log(ub / lb);
      assert q0 == q0 : "q0 is not a number " + q1;
      assert q1 == q1 : "q1 is not a number " + q1;
      assert Double.MIN_VALUE <= q0 && q0 <= Double.MAX_VALUE : "|q0| is too big" + Math.abs(q0);
      assert Double.MIN_VALUE <= q1 && q1 <= Double.MAX_VALUE : "|q1| is too big" + Math.abs(q1);
      assert logL + q0 > logL : "logL (" + logL + ") is too much smaller than q0 (" + q0 + ")";
      assert logU - q1 < logU : "logU (" + logU + ") is too much bigger than q1 (" + q1 + ")";
    }

    /**
     * @return {@link CompResult.Binary#LARGER}, {@link CompResult.Binary#SMALLER}, or {@link CompResult.Binary#UNDECIDED}.
     * <ul>
     * <li>{@link CompResult.Binary#LARGER} means that the test rejected the null hypothesis (the actual probability is not
     * smaller than the threshold in H0).</li>
     * <li>{@link CompResult.Binary#SMALLER} means that the test did not reject the null hypothesis.</li>
     * <li>{@link CompResult.Binary#UNDECIDED} means that the test is not done yet.</li>
     * </ul>
     * @ensures Let {@literal p} be the actual probability. There are two two probabilistic guarantees:
     * <ul>
     * <li>If {@code p <= threshold - delta} then the probability of returning {@link CompResult.Binary#LARGER} is at most {@link #alpha}.</li>
     * <li>If {@code p >= threshold + delta} then the probability of returning {@link CompResult.Binary#SMALLER} is at most {@link #beta}.</li>
     * </ul>
     */
    public CompResult.Binary status() {
      return logT <= logL ? CompResult.Binary.SMALLER :
             logT >= logU ? CompResult.Binary.LARGER :
             CompResult.Binary.UNDECIDED;
    }

    /**
     * Update the test by adding an observation
     * @param passed Whether or not the sample satisfied the test or not.
     */
    @Override
    public void update(final boolean passed) {
      if (passed) logT += q1;
      else logT += q0;
    }

    /**
     * Update the test by adding the input number of observations
     * @param passed Number of samples that are passed
     * @param failed Number of samples that are failed
     * @requires {@code passed >= 0}
     * @requires {@code failed >= 0}
     */
    @Override
    public void update(final int passed, final int failed) {
      assert passed >= 0 : "Invalid number of passed tests " + passed;
      assert failed >= 0 : "Invalid number of failed tests " + failed;
      logT += passed * q1 + failed * q0;
    }

    /** Reset the state (can be used to restart the test) */
    @Override
    public void reset() { logT = 0; }

    @Override
    public boolean completed() { return status() != CompResult.Binary.UNDECIDED; }

    @Override
    public boolean decided() { return true; }

    @Override
    public boolean nullHypRejected() { return status() == CompResult.Binary.LARGER; }

    @Override
    public String parametersStr() {
      return "threshold: " + threshold + ", " +
             "alpha: " + alpha + ", " +
             "beta: " + beta + ", " +
             "delta: " + delta +
             "logL: " + logL + ", " +
             "logU: " + logU + ", " +
             "q0: " + q0 + ", " +
             "q1: " + q1;
    }

    @Override
    public Binary cloneCopy() {
      try {
        return (Binary) super.clone();
      } catch (CloneNotSupportedException e) {
        throw new Error(e);
      }
    }

    /** Input Threshold */
    public final double threshold;

    /** Type I Error Probability */
    public final double alpha;

    /** Type II Error Probability */
    public final double beta;

    /** Half of the size of indifference region */
    public final double delta;

    /** Lower bound after which H0 won't be rejected */
    public final double logL;

    /** Upper bound after which H0 will be rejected */
    public final double logU;

    /** Step size for negative samples */
    public final double q0;

    /** Step size for positive samples */
    public final double q1;

    /** Test statistic */
    private double logT = 0;
  }

  /**
   * SPRT for ternary simple hypothesis testing.
   * <ul>
   * <li><em>Ternary</em> means if the algorithm terminates, there are three possibilities for its answer.
   * <li><em>Simple</em> means null and alternative hypotheses must be in the form of {@literal H0: p=p0} and
   * {@literal H1: p=p1}, respectively (for some values of {@code 0<p0<p1<1}). Note that {@literal H0: p<p0} is
   * not a simple hypothesis.</li>
   * </ul>
   */
  public static final class Ternary implements HypTest {

    /** Constructor used to cloning */
    private Ternary(Binary lbBinary, Binary ubBinary, double threshold, double alpha, double beta, double gamma, double delta) {
      this.lbBinary = lbBinary.cloneCopy();
      this.ubBinary = ubBinary.cloneCopy();
      this.threshold = threshold;
      this.alpha = alpha;
      this.beta = beta;
      this.gamma = gamma;
      this.delta = delta;
    }

    /**
     * Construct an instance of the ternary version of SPRT
     * @param threshold Input threshold
     * @param alpha     Type I error probability (also known as <em>false positive</em> probability) is the probability
     *                  of incorrectly rejecting the null hypothesis.
     * @param beta      Type II error probability (also known as <em>false negative</em> probability) is the probability
     *                  of incorrectly not rejecting the null hypothesis.
     * @param gamma     The probability of returning {@literal unknown} when the actual probability is at least
     *                  {@link #delta} away from the input threshold.
     * @param delta     Half of the size of indifference region.
     * @requires {@code 0 < lb < threshold < ub < ub < 1}
     * @requires {@code 0 < alpha < 0.5}
     * @requires {@code 0 < beta < 0.5}
     * @requires {@code 0 < gamma < 0.5}
     */
    public Ternary(double threshold, double alpha, double beta, double gamma, double delta) {
      assert 0 < threshold && threshold < 1 : "Wrong threshold " + threshold;
      assert 0 < alpha && alpha < 0.5 : "Wrong type I error " + alpha;
      assert 0 < beta && beta < 0.5 : "Wrong type II error " + beta;
      assert delta < threshold && threshold < 1 - delta : "Invalid threshold (" + threshold + ") and/or delta (" + delta + ")";
      assert 0 < gamma && gamma < 0.5 : "Wrong type III error " + gamma;
      this.threshold = threshold;
      this.alpha = alpha;
      this.beta = beta;
      this.delta = delta;
      this.gamma = gamma;
      final double halfDelta = delta / 2;
      this.lbBinary = new Binary(threshold - halfDelta, alpha, gamma, halfDelta);
      this.ubBinary = new Binary(threshold + halfDelta, gamma, beta, halfDelta);
    }

    /**
     * @param lb Result of lower-bound binary test
     * @param ub Result of upper-bound binary test
     * @return {@link CompResult.Ternary#LARGER}, {@link CompResult.Ternary#SMALLER}, {@link CompResult.Ternary#TOO_CLOSE}, or {@link CompResult.Ternary#UNDECIDED}.
     * <ul>
     * <li>{@link CompResult.Ternary#LARGER} means that the test rejected the null hypothesis (the actual probability is not
     * smaller than the threshold in H0).</li>
     * <li>{@link CompResult.Ternary#SMALLER} means that the test did not reject the null hypothesis.</li>
     * <li>{@link CompResult.Ternary#TOO_CLOSE} means that the test is finished, but could not determine an answer.</li>
     * <li>{@link CompResult.Ternary#UNDECIDED} means that the test is not done yet.</li>
     * </ul>
     * @ensures Let {@literal p} be the actual probability. There are two two probabilistic guarantees:
     * <ul>
     * <li>If {@code p <= threshold - delta} then the probability of returning {@link CompResult.Ternary#LARGER} is at most {@link #alpha}.</li>
     * <li>If {@code p >= threshold + delta} then the probability of returning {@link CompResult.Ternary#SMALLER} is at most {@link #beta}.</li>
     * <li>If {@code threshold - delta < p < threshold + delta} then the probability of returning {@link CompResult.Ternary#TOO_CLOSE} is at
     * most {@link #beta}.</li>
     * </ul>
     */
    public static CompResult.Ternary status(final CompResult.Binary lb, final CompResult.Binary ub) {
      return lb == CompResult.Binary.SMALLER ? CompResult.Ternary.SMALLER :
             ub == CompResult.Binary.LARGER ? CompResult.Ternary.LARGER :
             lb == CompResult.Binary.UNDECIDED || ub == CompResult.Binary.UNDECIDED ? CompResult.Ternary.UNDECIDED :
             CompResult.Ternary.TOO_CLOSE;
    }

    /**
     * Same as {@link #status(CompResult.Binary, CompResult.Binary)}, but input parameters are taken from the current state
     * @return Current status of the test
     */
    public CompResult.Ternary status() { return status(lbBinary.status(), ubBinary.status()); }

    /**
     * Update the test by adding an observation
     * @param passed Whether or not the sample satisfied the test or not.
     */
    @Override
    public void update(final boolean passed) {
      lbBinary.update(passed);
      ubBinary.update(passed);
    }

    /**
     * Update the test by adding an observation
     * @param passed Number of samples that are passed
     * @param failed Number of samples that are failed
     */
    @Override
    public void update(final int passed, final int failed) {
      lbBinary.update(passed, failed);
      ubBinary.update(passed, failed);
    }

    /** Reset the state (can be used to restart the test) */
    @Override
    public void reset() {
      lbBinary.reset();
      ubBinary.reset();
    }

    @Override
    public boolean completed() { return status() != CompResult.Ternary.UNDECIDED; }

    @Override
    public boolean decided() { return status() != CompResult.Ternary.TOO_CLOSE; }

    @Override
    public boolean nullHypRejected() { return status() == CompResult.Ternary.LARGER; }

    @Override
    public String parametersStr() {
      return "threshold: " + threshold + ", " +
             "alpha: " + alpha + ", " +
             "beta: " + beta + ", " +
             "gamma: " + gamma + ", " +
             "delta: " + delta + ", " +
             "lbBinary: " + lbBinary.parametersStr() + ", " +
             "ubBinary: " + ubBinary.parametersStr();
    }

    @Override
    public Ternary cloneCopy() { return new Ternary(lbBinary, ubBinary, threshold, alpha, beta, gamma, delta); }

    /** Input Threshold */
    public final double threshold;

    /** Type I Error Probability */
    public final double alpha;

    /** Type II Error Probability */
    public final double beta;

    /** The probability of incorrectly returning {@link CompResult.Ternary#TOO_CLOSE} */
    public final double gamma;

    /** Half of the size of indifference region */
    public final double delta;

    /** Lower bound binary test */
    private final Binary lbBinary;

    /** Upper bound binary test */
    private final Binary ubBinary;
  }
}

File: src/edu/stmc/SPRT.scala
package edu.stmc


// @formatter:off
/** Sequential Probability Ratio Test
  *
  * @note Probabilistic guarantees in this class ignore numerical errors caused by floating point arithmetic.
  * @constructor Construct a hypothesis test in which the null hypothesis is `p = t-δ` and the alternative hypothesis
  *              is `p = t+δ`, where `p` is the actual probability, `t` is the input threshold, and `δ` is the half of
  *              the size of indifference region.
  * @param threshold Input Threshold
  * @param alpha     Type I  Error Probability (also known as 'false positive' probability).
  * @param beta      Type II Error Probability (also known as 'false negative' probability).
  * @param delta     Half of the size of indifference region.
  * <pre>
  * The following requirements must be met by the main constructor:
  *   - 0 < threshold < 1
  *   - 0 < alpha < 0.5
  *   - 0 < beta < 0.5
  *   - 0 < delta < 0.5
  *   - delta < threshold
  *   - delta < 1 - threshold
  * </pre> */
// @formatter:on
class SPRT(private[this] val threshold: Double,
           private[this] val alpha: Double,
           private[this] val beta: Double,
           private[this] val delta: Double)
  extends HypTest {

  require(0 < threshold && threshold < 1, s"Invalid threshold $threshold")
  require(0 < alpha && alpha < 0.5, s"Invalid type I error $alpha")
  require(0 < beta && beta < 0.5, s"Invalid type II error $beta")
  require(0 < delta && delta < 0.5, s"Invalid indifference parameter $delta")
  require(delta < threshold && threshold < 1 - delta, s"Invalid threshold ($threshold) and/or delta ($delta)")
  private[this] val lb = threshold - delta
  private[this] val ub = threshold + delta

  /** Lower and upper bounds after which the null hypothesis won't/will be rejected */
  private[this] val (logL, logU) = (Math.log(beta / (1 - alpha)), Math.log((1 - beta) / alpha))
  assert(logL < 0, s"Lower-bound log is not negative " + logL)
  assert(logU > 0, s"Upper-bound log is not positive " + logU)

  /** Step sizes for negative/positive samples */
  private[this] val (q0, q1) = (Math.log((1 - ub) / (1 - lb)), Math.log(ub / lb))
  assert(java.lang.Double.isFinite(q0), s"q0 ($q0) is not a finite number")
  assert(java.lang.Double.isFinite(q1), s"q1 ($q1) is not a finite number")
  assert(logL - q0 > logL, s"logL ($logL) is too much smaller than q0 ($q0)")
  assert(logU - q1 < logU, s"logU ($logU) is too much bigger than q1 ($q1)")

  /** Test statistic */
  private[this] var logT = 0.0

  override def reset(): Unit = logT = 0

  override def update(passed: Boolean): Unit = if (passed) logT += q1 else logT += q0
  override def update(passed: Int, failed: Int): Unit = logT += passed * q1 + failed * q0

  /** @note The following probabilistic guarantees are made:
    *   1. If the actual probability is at most `threshold-delta` then the probability of returning
    *       [[CompResult.Binary.LARGER]] is at most `alpha`.
    *   1. If the actual probability is at least `threshold+delta` then the probability of returning
    *       [[CompResult.Binary.SMALLER]] is at most `beta`. */
  def status(logT: Double): CompResult.Binary =
    if (logT <= logL) CompResult.Binary.SMALLER
    else if (logT >= logU) CompResult.Binary.LARGER
    else CompResult.Binary.UNDECIDED

  /** Same as [[status(logT*]], but input parameter is taken from the current instance */
  @inline
  final def status: CompResult.Binary = status(logT)

  override def completed: Boolean = status ne CompResult.Binary.UNDECIDED

  /** @note [[completed]] does not need to be `true` (as required by [[HypTest]]). */
  override def decided: Boolean = status ne CompResult.Binary.UNDECIDED

  /** @note Requires [[completed]] to be `true`.
    * @note If the actual probability is at most `threshold-delta` then the probability of returning `true` is at most
    *       `alpha`. */
  override def rejected: Boolean = status eq CompResult.Binary.LARGER

  /** @note Requires [[completed]] to be `true`.
    * @note If the actual probability is at least `threshold+delta` then the probability of returning `true` is at most
    *       `beta`. */
  override def failed_to_reject: Boolean = status eq CompResult.Binary.LARGER

  override def parametersStr: String =
    s"threshold: $threshold, alpha: $alpha, beta: $beta, delta: $delta, logL: $logL, logU: $logU, q0: $q0, q1: $q1"

  override def toString: String = s"$parametersStr, logT: $logT"

  override def copy(): SPRT = clone.asInstanceOf[SPRT]
}

object SPRT {
  def apply(threshold: Double, alpha: Double, beta: Double, delta: Double) =
    new SPRT(threshold, alpha, beta, delta)
}
